### ISSUE TITLE
configure error (ld: unknown option)

### ISSUE THREAD

**Issue by chunhungyeh** [MRtrix3/mrtrix3#1243]
I am having a configure issue on the current ``dev`` branch and think it is related to the changes in #1143 (sorry for a late feedback for those changes!). Here are the output messages of running ``./configure`` from my macOS High Sierra (10.13.3) system:

The configure.log for the release mode - 
```

REPORT: 
MRtrix build type requested:

REPORT: release

REPORT: 

REPORT: Detecting OS: darwin
EXEC <<
CMD: sw_vers -productVersion
EXIT: 0
STDOUT:
10.13.3
>>


REPORT: OS X deployment target: 10.13

REPORT: Looking for compiler [clang++]:
EXEC <<
CMD: clang++ --version
EXIT: 0
STDOUT:
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin17.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
>>


REPORT: Apple LLVM version 9.0.0 (clang-900.0.39.2)

REPORT: Checking for C++11 compliance:

COMPILE /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpCaXBX8.cpp:
---

#include <cstddef>
struct Base {
    Base (int);
};
struct Derived : Base {
    using Base::Base;
};

int main() {
  Derived D (int); // check for contructor inheritance
  return 0;
}

---
EXEC <<
CMD: clang++ -c -std=c++11 -DMRTRIX_MACOSX -fPIC -mmacosx-version-min=10.13 /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpCaXBX8.cpp -o /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpCaXBX8.o
EXIT: 0
>>

EXEC <<
CMD: clang++ /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpCaXBX8.o -Wl,-O1,--sort-common,--as-needed -mmacosx-version-min=10.13 -o a.out
EXIT: 1
STDERR:
ld: unknown option: -O1
clang: error: linker command failed with exit code 1 (use -v to see invocation)
>>

error deleting temporary file "a.out": No such file or directory
REPORT: test failed (see configure.log for details)

REPORT: Looking for compiler [g++]:
EXEC <<
CMD: g++ --version
EXIT: 0
STDOUT:
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin17.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
STDERR:
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
>>


REPORT: Apple LLVM version 9.0.0 (clang-900.0.39.2)

REPORT: Checking for C++11 compliance:

COMPILE /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmp6DUDMR.cpp:
---

#include <cstddef>
struct Base {
    Base (int);
};
struct Derived : Base {
    using Base::Base;
};

int main() {
  Derived D (int); // check for contructor inheritance
  return 0;
}

---
EXEC <<
CMD: g++ -c -std=c++11 -DMRTRIX_MACOSX -fPIC -mmacosx-version-min=10.13 /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmp6DUDMR.cpp -o /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmp6DUDMR.o
EXIT: 0
>>

EXEC <<
CMD: g++ /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmp6DUDMR.o -Wl,-O1,--sort-common,--as-needed -mmacosx-version-min=10.13 -o a.out
EXIT: 1
STDERR:
ld: unknown option: -O1
clang: error: linker command failed with exit code 1 (use -v to see invocation)
>>

error deleting temporary file "a.out": No such file or directory
REPORT: test failed (see configure.log for details)

ERROR: no suitable compiler found!



  Set the CXX environment variable to inform 'configure' of the path to the
  compiler on your system, as follows:
    $ export CXX=/usr/bin/g++-5.5
    $./configure
  (amend with the actual path to the compiler on your system)

  If you are using a compiler other than gcc or clang, you can also set the CXX_ARGS
  environment variable to specify how your compiler expects different arguments
  to be presented on the command line, for instance as follows:
    $ export CXX_ARGS="-c CFLAGS SRC -o OBJECT"
    $ ./configure


  See the file 'configure.log' for details. If this doesn't help and you need
  further assistance, please post on the MRtrix3 community forum
  (http://community.mrtrix.org/), and make sure to include the full contents of
  the 'configure.log' file.
```

The configure.log for the debug mode - 
```

REPORT: 
MRtrix build type requested:

REPORT: debug

REPORT: 

REPORT: Detecting OS: darwin
EXEC <<
CMD: sw_vers -productVersion
EXIT: 0
STDOUT:
10.13.3
>>


REPORT: OS X deployment target: 10.13

REPORT: Looking for compiler [clang++]:
EXEC <<
CMD: clang++ --version
EXIT: 0
STDOUT:
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin17.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
>>


REPORT: Apple LLVM version 9.0.0 (clang-900.0.39.2)

REPORT: Checking for C++11 compliance:

COMPILE /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpsA9aNZ.cpp:
---

#include <cstddef>
struct Base {
    Base (int);
};
struct Derived : Base {
    using Base::Base;
};

int main() {
  Derived D (int); // check for contructor inheritance
  return 0;
}

---
EXEC <<
CMD: clang++ -c -std=c++11 -DMRTRIX_MACOSX -fPIC -mmacosx-version-min=10.13 /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpsA9aNZ.cpp -o /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpsA9aNZ.o
EXIT: 0
>>

EXEC <<
CMD: clang++ /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpsA9aNZ.o -Wl,-O0,--sort-common,--as-needed -mmacosx-version-min=10.13 -o a.out
EXIT: 1
STDERR:
ld: unknown option: -O0
clang: error: linker command failed with exit code 1 (use -v to see invocation)
>>

error deleting temporary file "a.out": No such file or directory
REPORT: test failed (see configure.log for details)

REPORT: Looking for compiler [g++]:
EXEC <<
CMD: g++ --version
EXIT: 0
STDOUT:
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin17.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
STDERR:
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
>>


REPORT: Apple LLVM version 9.0.0 (clang-900.0.39.2)

REPORT: Checking for C++11 compliance:

COMPILE /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpjYN51x.cpp:
---

#include <cstddef>
struct Base {
    Base (int);
};
struct Derived : Base {
    using Base::Base;
};

int main() {
  Derived D (int); // check for contructor inheritance
  return 0;
}

---
EXEC <<
CMD: g++ -c -std=c++11 -DMRTRIX_MACOSX -fPIC -mmacosx-version-min=10.13 /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpjYN51x.cpp -o /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpjYN51x.o
EXIT: 0
>>

EXEC <<
CMD: g++ /var/folders/j7/pqtdjgfj0kb0rfl1kmz08njh0000gn/T/tmpjYN51x.o -Wl,-O0,--sort-common,--as-needed -mmacosx-version-min=10.13 -o a.out
EXIT: 1
STDERR:
ld: unknown option: -O0
clang: error: linker command failed with exit code 1 (use -v to see invocation)
>>

error deleting temporary file "a.out": No such file or directory
REPORT: test failed (see configure.log for details)

ERROR: no suitable compiler found!



  Set the CXX environment variable to inform 'configure' of the path to the
  compiler on your system, as follows:
    $ export CXX=/usr/bin/g++-5.5
    $./configure
  (amend with the actual path to the compiler on your system)

  If you are using a compiler other than gcc or clang, you can also set the CXX_ARGS
  environment variable to specify how your compiler expects different arguments
  to be presented on the command line, for instance as follows:
    $ export CXX_ARGS="-c CFLAGS SRC -o OBJECT"
    $ ./configure


  See the file 'configure.log' for details. If this doesn't help and you need
  further assistance, please post on the MRtrix3 community forum
  (http://community.mrtrix.org/), and make sure to include the full contents of
  the 'configure.log' file.
```

Cheers.

**Comment by Lestropie** [comment]
OK, that's a bit sticky. This works happily with Clang on Linux, and the executable name is the same (and your "`g++`" is just a softlink to the same target executable as `clang++`).

I don't see a clean way of automagically parsing from the linker help pages whether or not the `-O#` command-line flag is available. So the options might be:

- To disable those linker optimisation flags outright on `darwin`. Assumes this is unsupported on all `darwin` platforms / compiler & linker versions.

- To run a test as part of `configure` to see if the relevant linker succeeds with the optimisation flag; if it doesn't, re-run the test with the optimisation flag removed. This is maybe the more general & robust solution.

@jdtournier: Would this affect production of release binaries for OSX as described in #1143?

**Comment by jdtournier** [comment]
I reckon we just leave the `-O1/0` flags out altogether - I'm not even sure what they do, to be honest... The main thing I was after was the `--as-needed` option, everything else was added because I'd seen it used elsewhere, particularly Qt. Let's just take it out...

**Comment by maxpietsch** [comment]
It is not the `-O1` flag specifically. Any flag in `-O1,--sort-common,--as-needed` passed to the linker triggers this on macOS:

```
clang++ test.o -Wl,--as-needed -mmacosx-version-min=10.13 -o a.out -v
Apple LLVM version 9.0.0 (clang-900.0.39.2)
Target: x86_64-apple-darwin17.3.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld" -demangle -lto_library /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libLTO.dylib -dynamic -arch x86_64 -macosx_version_min 10.13.0 -o a.out test.o --as-needed -lc++ -lSystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/9.0.0/lib/darwin/libclang_rt.osx.a
ld: unknown option: --as-needed
```

As a side note:

If we check linker flags compatibility in a `configure` test, I'd suggest also adding one for the [thin LTO](https://clang.llvm.org/docs/ThinLTO.html) option for clang (>=3.9) and use that for non-debug builds to save memory.

**Comment by jdtournier** [comment]
> If we check linker flags compatibility in a configure test, I'd suggest also adding one for the thin LTO option for clang (>=3.9) and use that for non-debug builds to save memory.

A couple of thoughts on that front: thin LTO (or any form of LTO) requires the corresponding flags to be passed to the compiler also - it needs to inject quite a bit more information into the compiled objects for the linker to be able to perform these optimisations. More to the point though, the benefits of LTO don't strike me as overwhelming... Looking at [this page](http://blog.llvm.org/2016/06/thinlto-scalable-and-incremental-lto.html), we're talking about performance gains of the order of ~3% on average - not something I'd go out of my way for, if I'm honest... 

Also not sure that flag would work on `gcc` on macOS, in case people are actually compiling it with homebrew's compiler...?

**Comment by Lestropie** [comment]
Closed by #1246.
